### PR TITLE
refactor: update `blas/ext/base/dsum` to follow current project conventions

### LINKED ISSUE #1504: [RFC]: refactor `blas/ext/base/dsum` to follow current project conventions
### Description

This RFC proposes refactoring [`@stdlib/blas/ext/base/dsum`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/blas/ext/base/dsum) to follow current project conventions.

These conventions are outlined in more detail in https://github.com/stdlib-js/stdlib/issues/788. Namely, we want to migrate from C++ add-on interfaces to C add-on interfaces and then make various style and simplification changes.

Some example commits which perform most, if not all, of the requested changes include https://github.com/stdlib-js/stdlib/commit/d66d80cfcbab1da95cb5ae8cfae4a1fc4c5129d1 (saxpy),  https://github.com/stdlib-js/stdlib/commit/3c82fe7b6c0f9fd6ef02c86d6e2024d708ac8c14 (cswap), and https://github.com/stdlib-js/stdlib/commit/5b15c9675be434c0e11b1ec1e1f78d7b089abbcd (dasum).

### Related Issues

Related issues #788, #1152.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by performant23**
Resolves #1504

## Description

> What is the purpose of this pull request?

This pull request:

- refactors the `blas/ext/base/dsum` to follow current project conventions i.e. C++ add-on interfaces to C add-on interfaces along with style and simplification changes.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #1504 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/blas/ext/base/dsum/README.md [hunk 1]
```diff
@@ -56,16 +56,15 @@ The function has the following parameters:
 -   **x**: input [`Float64Array`][@stdlib/array/float64].
 -   **stride**: index increment for `x`.
 
-The `N` and `stride` parameters determine which elements in `x` are accessed at runtime. For example, to compute the sum of every other element in `x`,
+The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to compute the sum of every other element in `x`,
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/README.md]:
```suggestion
The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to compute the sum of every other element in the strided array,
```

File: lib/node_modules/@stdlib/blas/ext/base/dsum/README.md [hunk 2]
```diff
@@ -56,16 +56,15 @@ The function has the following parameters:
...
 
 ```javascript
 var Float64Array = require( '@stdlib/array/float64' );
-var floor = require( '@stdlib/math/base/special/floor' );
 
 var x = new Float64Array( [ 1.0, 2.0, 2.0, -7.0, -2.0, 3.0, 4.0, 2.0 ] );
-var N = floor( x.length / 2 );
+var N = 4;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/blas/ext/base/dsum/README.md [hunk 3]
```diff
@@ -75,15 +74,14 @@ Note that indexing is relative to the first index. To introduce an offset, use [
 
 ```javascript
 var Float64Array = require( '@stdlib/array/float64' );
-var floor = require( '@stdlib/math/base/special/floor' );
 
 var x0 = new Float64Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
 var x1 = new Float64Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
 
-var N = floor( x0.length / 2 );
+var N = 3;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/blas/ext/base/dsum/README.md [hunk 4]
```diff
@@ -94,26 +92,25 @@ Computes the sum of double-precision floating-point strided array elements using
 var Float64Array = require( '@stdlib/array/float64' );
 
 var x = new Float64Array( [ 1.0, -2.0, 2.0 ] );
-var N = x.length;
+var N = 3;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/blas/ext/base/dsum/README.md [hunk 5]
```diff
@@ -75,15 +73,12 @@ Note that indexing is relative to the first index. To introduce an offset, use [
...
-var floor = require( '@stdlib/math/base/special/floor' );
 
 var x0 = new Float64Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
 var x1 = new Float64Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
 
-var N = floor( x0.length / 2 );
-
-var v = dsum( N, x1, 2 );
-// returns 5.0
+var v = dsum( 3, x1, 2 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/README.md]:
```suggestion
var v = dsum( 4, x1, 2 );
```

File: lib/node_modules/@stdlib/blas/ext/base/dsum/README.md [hunk 6]
```diff
@@ -94,26 +89,24 @@ Computes the sum of double-precision floating-point strided array elements using
...
+var v = dsum.ndarray( 3, x, 1, 0 );
 // returns 1.0
 ```
 
 The function has the following additional parameters:
 
 -   **offset**: starting index for `x`.
 
-While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on a starting index. For example, to calculate the sum of every other value in `x` starting from the second value
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the offset parameter supports indexing semantics based on a starting index. For example, to calculate the sum of every other value in `x` starting from the second value
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/README.md]:
```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the offset parameter supports indexing semantics based on a starting index. For example, to calculate the sum of every other value in the strided array starting from the second value
```

File: lib/node_modules/@stdlib/blas/ext/base/dsum/README.md [hunk 7]
```diff
@@ -94,26 +89,24 @@ Computes the sum of double-precision floating-point strided array elements using
...
 
 ```javascript
 var Float64Array = require( '@stdlib/array/float64' );
-var floor = require( '@stdlib/math/base/special/floor' );
 
 var x = new Float64Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
-var N = floor( x.length / 2 );
+var N = 4;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/blas/ext/base/dsum/docs/repl.txt [hunk 8]
```diff
@@ -33,21 +33,20 @@
     > {{alias}}( x.length, x, 1 )
     1.0
 
-    // Using `N` and `stride` parameters:
+    // Using `N` and stride parameters:
     > x = new {{alias:@stdlib/array/float64}}( [ -2.0, 1.0, 1.0, -5.0, 2.0, -1.0 ] );
-    > var N = {{alias:@stdlib/math/base/special/floor}}( x.length / 2 );
     > var stride = 2;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/docs/repl.txt]:
```suggestion
```

File: lib/node_modules/@stdlib/blas/ext/base/dsum/docs/repl.txt [hunk 9]
```diff
@@ -33,21 +33,20 @@
...
-    > {{alias}}( N, x, stride )
+    > {{alias}}( 3, x, stride )
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/docs/repl.txt]:
```suggestion
    > {{alias}}( 3, x, 2 )
```

File: lib/node_modules/@stdlib/blas/ext/base/dsum/docs/repl.txt [hunk 10]
```diff
@@ -33,21 +33,20 @@
...
     1.0
 
     // Using view offsets:
     > var x0 = new {{alias:@stdlib/array/float64}}( [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0 ] );
     > var x1 = new {{alias:@stdlib/array/float64}}( x0.buffer, x0.BYTES_PER_ELEMENT*1 );
-    > N = {{alias:@stdlib/math/base/special/floor}}( x0.length / 2 );
     > stride = 2;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/docs/repl.txt]:
```suggestion
```

File: lib/node_modules/@stdlib/blas/ext/base/dsum/docs/repl.txt [hunk 11]
```diff
@@ -33,21 +33,20 @@
...
-    > {{alias}}( N, x1, stride )
+    > {{alias}}( 3, x1, stride )
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/docs/repl.txt]:
```suggestion
    > {{alias}}( 3, x1, 2 )
```

File: lib/node_modules/@stdlib/blas/ext/base/dsum/docs/repl.txt [hunk 12]
```diff
@@ -84,10 +83,8 @@
 
     // Using offset parameter:
     > var x = new {{alias:@stdlib/array/float64}}( [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0 ] );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/docs/repl.txt]:
```suggestion
    > x = new {{alias:@stdlib/array/float64}}( [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0 ] );
```

File: lib/node_modules/@stdlib/blas/ext/base/dsum/docs/repl.txt [hunk 13]
```diff
@@ -84,10 +83,8 @@
...
-    > var N = {{alias:@stdlib/math/base/special/floor}}( x.length / 2 );
-    > {{alias}}.ndarray( N, x, 2, 1 )
+    > {{alias}}.ndarray( 3, x, 2, 1 )
     -1.0
 
     See Also
     --------
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/docs/repl.txt]:
```suggestion
    --------

```

File: lib/node_modules/@stdlib/blas/ext/base/dsum/manifest.json [hunk 14]
```diff
@@ -35,7 +35,12 @@
 			],
 			"libpath": [],
 			"dependencies": [
-				"@stdlib/blas/ext/base/dsumkbn"
+				"@stdlib/blas/ext/base/dsumkbn",
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/manifest.json]: @performant23 Would you mind updating this file to use 2-space indentation? Thanks!

File: lib/node_modules/@stdlib/blas/ext/base/dsum/src/addon.c [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+* @private
+* @param env    environment under which the function is invoked
+* @param info   callback data
+* @return       Node-API value
+*/
+static napi_value addon( napi_env env, napi_callback_info info ) {
+    STDLIB_NAPI_ARGV( env, info, argv, argc, 3 );
+    STDLIB_NAPI_ARGV_INT64( env, N, argv, 0 );
+    STDLIB_NAPI_ARGV_STRIDED_FLOAT64ARRAY( env, X, N, stride, argv, 1 );
+    STDLIB_NAPI_ARGV_INT64( env, stride, argv, 2 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/src/addon.c]:
```suggestion
	STDLIB_NAPI_ARGV( env, info, argv, argc, 3 );
	STDLIB_NAPI_ARGV_INT64( env, N, argv, 0 );
	STDLIB_NAPI_ARGV_INT64( env, stride, argv, 2 );
	STDLIB_NAPI_ARGV_STRIDED_FLOAT64ARRAY( env, X, N, stride, argv, 1 );
```

You need to parse the stride before the strided array.

File: lib/node_modules/@stdlib/blas/ext/base/dsum/src/addon.c [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+
+	napi_value v;
+	napi_status status = napi_create_double(env, (double)stdlib_strided_dsum( N, (double *)X, stride ), &v );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/src/addon.c]:
```suggestion
	napi_status status = napi_create_double( env, (double)stdlib_strided_dsum( N, (double *)X, stride ), &v );
```

**Review by kgryte — CHANGES_REQUESTED** [review]
Apart from updating the indentation of `manifest.json`, this PR is looking good.

File: lib/node_modules/@stdlib/blas/ext/base/dsum/README.md [hunk 17]
```diff
@@ -75,15 +73,12 @@ Note that indexing is relative to the first index. To introduce an offset, use [
...
 
 var x0 = new Float64Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
 var x1 = new Float64Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
 
-var N = floor( x0.length / 2 );
-
-var v = dsum( N, x1, 2 );
-// returns 5.0
+var v = dsum( 4, x1, 2 );
+// returns 1.0
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/README.md]:
```suggestion
// returns 5.0
```

Comment (performant23) [reply lib/node_modules/@stdlib/blas/ext/base/dsum/README.md]: Hi, was just checking the error log, thanks for the change!

File: lib/node_modules/@stdlib/blas/ext/base/dsum/manifest.json (hunk 14 above)
Comment (performant23) [reply lib/node_modules/@stdlib/blas/ext/base/dsum/manifest.json]: Hello, @kgryte, thanks for your feedback! I checked the file here and on my local environment (with configuring for whitespace changes with 'git config core.whitespace cr-at-eol' and using 'git diff' and its variants) but both didn't detect any changes. Would you mind checking, please? It's possible that the issue might be related to local environment configurations. I've noticed that all the .json files default to 4 spaces on my setup, as configured that way.

Comment (kgryte) [reply lib/node_modules/@stdlib/blas/ext/base/dsum/manifest.json]: You should be using EditorConfig, as mentioned in the dev guide. We have per file type configuration. See the EditorConfig file in the root project directory.

Comment (performant23) [reply lib/node_modules/@stdlib/blas/ext/base/dsum/manifest.json]: Yeah, right! Thanks @kgryte! Have made the changes!

Comment (kgryte) [reply lib/node_modules/@stdlib/blas/ext/base/dsum/manifest.json]: Awesome. Thanks! Will review again later today.

File: lib/node_modules/@stdlib/blas/ext/base/dsum/manifest.json [hunk 18]
```diff
@@ -1,42 +1,47 @@
...
-			"libraries": [
-				"-lm"
-			],
-			"libpath": [],
-			"dependencies": [
-				"@stdlib/blas/ext/base/dsumkbn"
-			]
-		}
-	]
+  "options": {},
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/manifest.json]: We also need to add additional configuration to this file. See, e.g., https://github.com/stdlib-js/stdlib/blob/cf29ed0fc4f7108b5be01daca89ef72e6ebfcc1f/lib/node_modules/%40stdlib/blas/ext/base/dfill/manifest.json.

File: lib/node_modules/@stdlib/blas/ext/base/dsum/src/addon.c [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "stdlib/blas/ext/base/dsum.h"
+#include "stdlib/napi/export.h"
+#include "stdlib/napi/argv.h"
+#include "stdlib/napi/argv_float.h"
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/src/addon.c]:
```suggestion
```

Comment (kgryte) [reply lib/node_modules/@stdlib/blas/ext/base/dsum/src/addon.c]: This header is not used.

File: lib/node_modules/@stdlib/blas/ext/base/dsum/manifest.json [hunk 20]
```diff
@@ -1,42 +1,47 @@
...
+      ],
+      "libraries": [
+        "-lm"
+      ],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/blas/ext/base/dsumkbn",
+        "@stdlib/napi/export",
+        "@stdlib/napi/argv",
+        "@stdlib/napi/argv-float",
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/manifest.json]:
```suggestion
```

This dep is not used.

File: lib/node_modules/@stdlib/blas/ext/base/dsum/lib/ndarray.native.js [hunk 21]
```diff
@@ -47,10 +48,11 @@ var addon = require( './dsum.native.js' );
...
 	var view;
-	if ( stride < 0 ) {
-		offset += (N-1) * stride;
-	}
-	view = new Float64Array( x.buffer, x.byteOffset+(x.BYTES_PER_ELEMENT*offset), x.length-offset ); // eslint-disable-line max-len
+
+	offset = minViewBufferIndex( N, stride, offset );
+
+	view = offsetView( x, offset );
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/lib/ndarray.native.js]:
```suggestion
	offset = minViewBufferIndex( N, stride, offset );
	view = offsetView( x, offset );
```

File: lib/node_modules/@stdlib/blas/ext/base/dsum/src/addon.c [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+* @return       Node-API value
+*/
+static napi_value addon( napi_env env, napi_callback_info info ) {
+	STDLIB_NAPI_ARGV( env, info, argv, argc, 3 );
+	STDLIB_NAPI_ARGV_INT64( env, N, argv, 0 );
+	STDLIB_NAPI_ARGV_INT64( env, stride, argv, 2 );
+	STDLIB_NAPI_ARGV_STRIDED_FLOAT64ARRAY( env, X, N, stride, argv, 1 );
+
+	napi_value v;
+	napi_status status = napi_create_double( env, (double)stdlib_strided_dsum( N, (double *)X, stride ), &v );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/src/addon.c]:
```suggestion
	napi_status status = napi_create_double( env, stdlib_strided_dsum( N, X, stride ), &v );
```

The explicit casts are not necessary.

**Review by kgryte — CHANGES_REQUESTED** [review]
After another review, the only thing I can see which needs to be updated is the inclusion of additional build configurations to `manifest.json`.

File: lib/node_modules/@stdlib/blas/ext/base/dsum/manifest.json [hunk 23]
```diff
@@ -1,42 +1,77 @@
...
+      "libraries": [
+        "-lm"
+      ],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/blas/ext/base/dsumkbn",
+        "@stdlib/napi/export",
+        "@stdlib/napi/argv",
+        "@stdlib/napi/argv-int64",
+        "stdlib/napi/argv-strided-float64array"
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dsum/manifest.json]:
```suggestion
        "@stdlib/napi/argv-strided-float64array"
```

**Review by kgryte — APPROVED** [review]
LGTM. Will merged once CI passes. Thanks, @performant23!

**Comment by performant23** [comment]
Hi @kgryte, Thanks for the thorough review and the changes since this one was a bit convoluted because it was made a while ago!

**Pull request merged**